### PR TITLE
[14.0] [IMP] add function to generate unique project key automatically.

### PR DESCRIPTION
**Pull request by yweng8111**
As we discussed in #966 , I have implemented a small extension for automatic generating of the unique project key.

### CODE REVIEW

**Review by marylla — APPROVED** [review]
I made a functional review.

I created 3 projects with similiar naming scheme, and tested it in 2 different runboat instances **with** and **without** this improvement and checked the key suggestions.

| Project Name              | Without Improvement | With Improvement |
| ---| ---| ---|
| "Internal Event"           | IE | IE |
| "International Event"   | IE | IE1 |
| "Intercompany Event" | IE | IE2 |

So, I guess, the problem of the unique contraint and the hook that fixes old projects (before installation of module project key) is fixed this way.

Thank you! :+1:

**Review by chandni299 — APPROVED** [review]
It looks good

File: project_key/models/project_project.py [hunk 1]
```diff
@@ -138,12 +138,29 @@ def generate_project_key(self, text):
...
+        counter = 0
+        while not unique_key:
+            if counter != 0:
+                res = "%s%s" % (text, counter)
+            projects = self.search([("key", "=", res)])
+            if not projects:
+                unique_key = True
+            else:
+                counter += 1
+
```
Comment (pedrobaeza) [inline project_key/models/project_project.py]:
```suggestion
            unique_key = bool(self.search([("key", "=", res)]))
            counter += 1
```

File: project_key/models/project_project.py [hunk 2]
```diff
@@ -138,12 +138,29 @@ def generate_project_key(self, text):
 
         data = text.split(" ")
         if len(data) == 1:
-            return data[0][:3].upper()
+            res = data[0][:3].upper()
+            return self._generate_project_unique_key(res)
```
Comment (pedrobaeza) [inline project_key/models/project_project.py]:
```suggestion
            return self._generate_project_unique_key(data[0][:3].upper())
```

**Review by pedrobaeza — COMMENTED** [review]
Make code shorter

File: project_key/models/project_project.py (hunk 1 above)
Comment (yweng8111) [reply project_key/models/project_project.py]: Thanks for the review.
I've changed it.

File: project_key/models/project_project.py (hunk 2 above)
Comment (yweng8111) [reply project_key/models/project_project.py]: Done

**Comment by marylla** [comment]
@pedrobaeza This was very fast. Awesome. :) Could you merge?

**Pull request merged**